netbox/dcim/models/device_components.py

```diff
@@ -139,22 +134,42 @@ def get_peer_port(termination, position=1, follow_circuits=False):
             raise LoopDetected()
         cable_history.append(self.cable)
 
-        far_end = self.cable.termination_b if self.cable.termination_a == self else self.cable.termination_a
-        path = [(self, self.cable, far_end)]
+        # Determine the other end of our cable (like termination B)
+        other_end = self.cable.termination_b if self._cabled_as_a.exists() else self.cable.termination_a
+        path = [(self, self.cable, other_end)]
 
-        peer_port, position = get_peer_port(far_end, position, follow_circuits)
-        if peer_port is None:
+        if isinstance(other_end, RearPort) and other_end.positions > 1:
+            # When we end up here we have reached a rear port with multiple front ports, but we don't know which front
+            # port to continue with. So this is the end of the line.
             return path
 
-        try:
-            next_segment = peer_port.trace(position, follow_circuits, cable_history)
-        except LoopDetected:
+        from circuits.models import CircuitTermination
+        if isinstance(other_end, CircuitTermination) and not follow_circuits:
+            # Not tracing the circuit, stop here
             return path
 
-        if next_segment is None:
-            return path + [(peer_port, None, None)]
+        # What comes after the other end? (like termination C)
+        other_end_peer_port = other_end.get_peer_port()
+        while other_end_peer_port:
+            path += other_end_peer_port.trace(follow_circuits=follow_circuits, cable_history=cable_history)
+
+            if isinstance(other_end, FrontPort) and \
+                    isinstance(other_end_peer_port, RearPort) and other_end_peer_port.positions > 1:
```

Comment:
`other_end_peer_port` in this case can _only_ be a RearPort, correct? Because we know `other_end` is a FrontPort?

Reply:
I included it to cover the case where `other_end_peer_port` is None, but I realise that that is impossible because a front port already has a non-nullable ForeignKey to its rear port. I'd prefer to leave the check in: it doesn't hurt, makes it explicit what type `other_end_peer_port` is, and may catch unforeseen situations if the data model changes in the future.

But you are right: currently `isinstance(other_end_peer_port, RearPort)` will always succeed